app/src/main/java/org/kiwix/kiwixmobile/wifi_hotspot/HotspotService.java

```diff
@@ -0,0 +1,200 @@
+import androidx.annotation.Nullable;
+import androidx.core.app.NotificationCompat;
+import org.kiwix.kiwixmobile.R;
+import org.kiwix.kiwixmobile.utils.Constants;
+import org.kiwix.kiwixmobile.utils.ServerUtils;
+import org.kiwix.kiwixmobile.webserver.ZimHostCallbacks;
+import org.kiwix.kiwixmobile.webserver.WebServerHelper;
+import org.kiwix.kiwixmobile.webserver.ZimHostActivity;
+
+import static org.kiwix.kiwixmobile.webserver.ZimHostActivity.SELECTED_ZIM_PATHS_KEY;
+
+/**
+ * HotspotService is used to add a foreground service for the wifi hotspot.
+ * Created by Adeel Zafar on 07/01/2019.
+ */
+
+public class HotspotService extends Service {
+
+  public static final String ACTION_TURN_ON_AFTER_O = "Turn_on_hotspot_after_oreo";
+  public static final String ACTION_TURN_OFF_AFTER_O = "Turn_off_hotspot_after_oreo";
+  public static final String ACTION_IS_HOTSPOT_ENABLED = "Is_hotspot_enabled";
+  public static final String ACTION_START_SERVER = "start_server";
+  public static final String ACTION_STOP_SERVER = "stop_server";
+
+  private static final int HOTSPOT_NOTIFICATION_ID = 666;
+  private static final String ACTION_STOP = "hotspot_stop";
+  private static final String TAG = "HotspotService";
+  private WifiHotspotManager hotspotManager;
+  private BroadcastReceiver stopReceiver;
+  private NotificationManager notificationManager;
+  private NotificationCompat.Builder builder;
+  private ZimHostCallbacks zimHostCallbacks;
+  private final IBinder serviceBinder = new HotspotBinder();
+  private WebServerHelper webServerHelper;
+
+  @Override public void onCreate() {
+
+    super.onCreate();
+
+    hotspotManager = new WifiHotspotManager(this);
+
+    stopReceiver = new BroadcastReceiver() {
+      @Override
+      public void onReceive(Context context, Intent intent) {
+        if (intent != null && intent.getAction().equals(ACTION_STOP)) {
+          stopHotspotAndDismissNotification();
+        }
+      }
+    };
+
+    registerReceiver(stopReceiver, new IntentFilter(ACTION_STOP));
+
+    webServerHelper = new WebServerHelper();
+
+    notificationManager = (NotificationManager) getSystemService(Context.NOTIFICATION_SERVICE);
+  }
+
+  @Override public int onStartCommand(@NonNull Intent intent, int flags, int startId) {
+    switch (intent.getAction()) {
+
+      case ACTION_IS_HOTSPOT_ENABLED:
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
+          zimHostCallbacks.onHotspotStateReceived(hotspotManager.checkHotspotState());
+        }
+        break;
+
+      case ACTION_TURN_ON_AFTER_O:
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
+          hotspotManager.turnOnHotspot(zimHostCallbacks);
+        }
+        break;
+
+      case ACTION_TURN_OFF_AFTER_O:
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
+          stopHotspotAndDismissNotification();
+        }
+        break;
+
+      case ACTION_START_SERVER:
+        if (!webServerHelper.startServerHelper(
```

Comment:
As a general rule you should flip logic like this because reading a negative involves more mental processing
```
if(webServerDidStart){
  startServer()
}
else{
  stopService()
}
```
Is much more legible. 

I am also starting to question if `webServerHelper` should even be in here. Good advice is to keep Services as small as possible and delegate the heavy lifting and logic to other classes because god knows services are already complicated enough

Reply:
Fixed. f0975b0d239a57c44d2f7640eefe4b590798724d